test/server/listener_manager_impl_test.cc

```diff
@@ -4767,6 +4767,88 @@ traffic_direction: INBOUND
   EXPECT_CALL(*listener_foo_update1, onDestroy());
 }
 
+TEST_F(ListenerManagerImplForInPlaceFilterChainUpdateTest, RemoveTheInplaceUpdatingListener) {
```

Comment:
> it seems that this test doesn't reproduce the original crash

That's true.

> What is this test covering?

This test case introduces a timer out-of-order sequence. This test verified that listener_manager accomplishes the listener updates `if worker threads are doing the right thing` even in this out-of-order scenario.

BTW: connection_handler_test verifies worker thread would do the right. The new connection_handler_test fails before the one-liner fix.

I am adding the e2e test since it is not that hard to reproduce

Reply:
Sorry I was too arrogant. I haven't found a reliable approach to post the listener removal and drain listener in the same poll cycle in the e2e scenario.

Is there an example injecting the SimulatedTimeSystem into the integration test created envoy?

Reply:
I think that an e2e test that reproduces the crash some fraction of the time would be sufficient.

If you want to force the two events to execute in the same iteration, one possible way of doing it is to:
- post a callback to workers which waits for an absl::Notification in order to stall the workers
- run the config reload that removes the filter chain and listeners
- call absl::Notification::Notify() to allow the workers to continue

post doesn't use timers, so I don't think that SimulatedTimeSystem would help you repo more reliably.

Reply:
> I think that an e2e test that reproduces the crash some fraction of the time would be sufficient.

I am updating a commit. It highly depends on the execution of the execution.

> If you want to force the two events to execute in the same iteration, one possible way of doing it is to:

I will add comment to the existing TcpListenerRemoveFilterChainCalledAfterListenerIsRemoved to explain how comes the each step.

> post doesn't use timers, so I don't think that SimulatedTimeSystem would help you repo more reliably.

Post doesn't use timers, but the source of the post is triggered by timer. I explored move backward the system time but the enableTimer(duration) relies on monotonicTimer and it is designed to be forward only. So yes, I am blocked by an enable first by expired later behavior.

Reply:
> > I think that an e2e test that reproduces the crash some fraction of the time would be sufficient.
> 
> I am updating a commit. It highly depends on the execution of the execution.
> 
> > If you want to force the two events to execute in the same iteration, one possible way of doing it is to:
> 
> I will add comment to the existing TcpListenerRemoveFilterChainCalledAfterListenerIsRemoved to explain how comes the each step.
> 
> > post doesn't use timers, so I don't think that SimulatedTimeSystem would help you repo more reliably.
> 
> Post doesn't use timers, but the source of the post is triggered by timer. I explored move backward the system time but the enableTimer(duration) relies on monotonicTimer and it is designed to be forward only. So yes, I am blocked by an enable first by expired later behavior.

Time does not move forward in an iteration of the event loop, so if you schedule 2 timers in the same loop iteration and give them the same duration both timers will trigger in the same loop iteration when they expire.  The order in which timers with the same duration trigger is non-deterministic.

Reply:
That's exactly the root cause of this listener bug. Now the connection handler works correctly under this non-deterministic scenario. 

I wish I can manipulate the time in e2e test using SimulatedTimeSystem to produce the out-of-order in a deterministic way, but I failed. So the integration test does not crash everytime using the buggy code.

Reply:
We don't need it to crash every time.  How often does it manage to repo the crash?

Reply:
It highly depends on the load pattern.

The first build and test are likely to trigger it. My impression is that the crash is 50%.
However, if I run the test using --runs_per_test. The crash rate drops down to <1%.

I didn't drill into it

Reply:
I just chatted with @lambdai and this is covered accurately in unit tests. @jmarantz is it really impossibly to have `SimulatedTimeSystem` fire timers for the same duration in a different order to that in which they are set? Is there a simple path to add such behavior?

If not, I think we can move ahead and land this PR as-is.

Reply:
https://github.com/envoyproxy/envoy/pull/11962 randomized execution of simulated time timers scheduled for the same time.  If we do schedule timers at the same time, we should see tests trigger each of the orders half the time, assuming that random seeds are picked for each of the runs.

If randomization is not working as expected, let's go fix it.